### ISSUE TITLE
Ambiguity between auto derived and sdtlib instances in Cats

### ISSUE THREAD

**Issue by joroKr21** [milessabin/kittens#62]
Here is an example with `Functor` and `Option`:
```scala
import cats.instances.all._
import cats.derived.functor._
cats.Functor[Option] // ambiguous implicits
```
One possibility would be to export an implicit with the least specific type possible. For `Functor` that would mean something like this:
```scala
implicit def derivedFunctor[F[_]](mk: MkFunctor[F]): Functor[F] = mk
```
Currently this doesn't work for HKTs because of scala/bug#9445. E.g.
```scala
implicitly[Show[Int] <:< (Show[A] forSome { type A })] // works
implicitly[Functor[Option] <:< (Functor[F] forSome { type F[_] })] // doesn't work
```
The existentials above are involved in the overloading resolution for generic methods which is the mechanism used for implicit resolution in Scala.

**Comment by kailuowang** [comment]
What I did with `Show` is that prioritize the existing instance over derived instance. 
See https://github.com/milessabin/kittens/blob/master/core/src/main/scala/cats/derived/show.scala#L30-L43
Basically when deriving for a new type, it will try to search for a `Show` instance first, and then if not found it will try search for `MkShow` instance, which means that it will use the autoderivation in `MkShow`'s companion object only when `Show` is not available. 
We could do it with other derivations too.

**Comment by joroKr21** [comment]
I mean that they clash when we `import cats.derived.show._`, i.e. when we want auto derivation:
```scala
Error:(15, 21) ambiguous implicit values:
 both method genericDerivedShowCoproduct in trait MkShow3 of type [A, R <: shapeless.Coproduct](implicit repr: shapeless.LabelledGeneric.Aux[A,R], implicit s: shapeless.Lazy[cats.derived.MkShow[R]])cats.derived.MkShow[A]
 and method catsStdShowForOption in trait OptionInstances of type [A](implicit A: cats.Show[A])cats.Show[Option[A]]
 match expected type cats.Show[Option[String]]
    assert(foo.show == printedFoo)
```

What you did for `MkShow` is make semiauto derivation deeper so that it can work on nested case classes. This makes sense, but it's orthogonal. We can have `MkShow` both be a value class and have semiauto derivation for nested case classes.

**Comment by kailuowang** [comment]
I apologize, I am not really following you. So with your change, does it do semiauto derivation without conflicting with the existing instance in scope? where are you proposing to place `implicit def deriveEmpty[A](implicit mk: MkEmpty[A]): Empty[A] = mk.value`?
Inside `MkEmpty` will you use `Empty` or `MkEmpty`? 
Maybe a more complete example with `MkEmpty` and usage would help.

**Comment by joroKr21** [comment]
Yes, sorry if I was unclear, here's an example with `Semigroup`:
```scala
trait SemigroupDerivation {
  implicit def deriveSemigroup[A](
    implicit mk: Lazy[MkSemigroup[A]]
  ): Semigroup[A] = mk.value.value
}

class MkSemigroup[A](val value: Semigroup[A]) extends AnyVal
object MkSemigroup {
  def apply[A](comb: (A, A) => A): MkSemigroup[A] =
    new MkSemigroup(new Semigroup[A] {
      def combine(x: A, y: A) = comb(x, y)
    })

  implicit val mkSemigroupHnil: MkSemigroup[HNil] =
    apply((_, _) => HNil)

  implicit def mkSemigroupHcons[H, T <: HList](
    implicit sgH: Lazy[Semigroup[H]], sgT: MkSemigroup[T]
  ): MkSemigroup[H :: T] = apply {
    case (h1 :: t1, h2 :: t2) =>
      sgH.value.combine(h1, h2) :: sgT.value.combine(t1, t2)
  }

  implicit def mkSemigroupGeneric[A, R](
    implicit gen: Generic.Aux[A, R], sgR: MkSemigroup[R]
  ): MkSemigroup[A] = apply { (x, y) =>
    gen.from(sgR.value.combine(gen.to(x), gen.to(y)))
  }
}
```
Then we can use automatic derivation without clashes:
```scala
import cats.instances.all._
import cats.derived.semigroup._
case class Bar(i: Int, ls: (Long, String))
cats.Semigroup[Bar] // not ambiguous anymore
```

**Comment by kailuowang** [comment]
Interesting, what I didn't get is that by making `MkTC` a value class wrapping `TC`, 
`implicit def deriveEmpty[A](implicit mk: MkEmpty[A]): Empty[A] = mk.value` somehow becomes least specific? While, if you just make `MkTC` inherit `TC`, then `implicit def deriveEmpty[A](implicit mk: MkEmpty[A]): Empty[A] = mk` has the same specificity as the other instance in scope, right? 

I am curious which implicit resolution rule leads to that result. But yeah if that solves the conflicts we should adopt it.

**Comment by joroKr21** [comment]
Hmm, actually it will be enough to define `MkSemigroupDerivation` like this:
```scala
trait MkSemigroupDerivation {
  implicit def deriveSemigroup[A](implicit mk: MkSemigroup[A]): Semigroup[A] = mk
}
```
The important part is not to import the `mkSemigroup` implicits. There are two specificity rules that are at odds with each other:
1. `MkSemigroup[A] <:< Semigroup[A]` and is therefore more specific.
2. `Semigroup[(A, B)]` is more specific than `Semigroup[A]`

**Comment by kailuowang** [comment]
if that works it effectively lower the priority of the derivation right? Will you be interested in a PR to change just `semigroup` or `show` first?

**Comment by kailuowang** [comment]
It looks like for some type class, this method has the same specificity as the imported instance. So I am not sure we can apply it across the board. 
If you want, you can try it with `Functor`

**Comment by joroKr21** [comment]
Yeah, it seems that specificity rules (or rather - subtyping) break down when it comes to HKT type parameters. Thanks very much, existentials:
```scala
trait Tc[A]
trait TcHk[F[_]]
implicitly[Tc[Int] <:< (Tc[A] forSome { type A })] // sure
implicitly[TcHk[Option] <:< (TcHk[F] forSome { type F[_] })] // nope
```

**Comment by kailuowang** [comment]
The semiautomatic derivation enabled in #63 is probably the best solution we can have for now .

**Comment by kailuowang** [comment]
Shall we close this one?

**Comment by kailuowang** [comment]
I am fine with keeping this one open, but my proposal is to disable auto derivation altogether and use semi-auto ones instead. A blanket auto derivation for everything in the file will always be kind of slippery in my view.

**Comment by joroKr21** [comment]
Fixed in Scala by scala/scala#6122 and scala/scala#6123. Also, Kittens now uses a more flexible scheme.
